### ISSUE TITLE
Wrong output dtype for nn.Linear under autocast and vmap

### ISSUE THREAD

**Issue by ValerianRey** [pytorch/pytorch#192617]
### 🐛 Describe the bug

Let's look at a Linear model under autocast:
```python
import torch

model = torch.nn.Linear(2, 3)
x = torch.ones(16, 2)

with torch.autocast("cpu", dtype=torch.float16):
    y = model(x)
   
print(y.dtype)
```
Expectedly, this outputs `torch.float16`.

Now, if we vmap the model:
```python
import torch

model = torch.nn.Linear(2, 3)
vmodel = torch.vmap(model)
x = torch.ones(16, 2)

with torch.autocast("cpu", dtype=torch.float16):
    y = vmodel(x)
   
print(y.dtype)
```
The output this time is `torch.float32`. I would expect these two examples to give the exact same result, but they differ in dtype. After some investigation, it seems that internally, under vmap, the matrix multiplication is done in fp16, but the addition of the bias term is done in fp32, and thus the result is in fp32. I could confirm this: running this second example with `bias=False` makes the result be, as expected, `torch.float16`.

I think fixing this would make another bug surface: the forward batching rules of the normalization layers (batch_norm, etc.) bug if their inputs are fp16, which this bug prevents from happening.
The same bug exists for their backward batching rules, reported in #184890 and that I'm trying to fix in #192161.

Also, given the fact I've found already 3 bugs in vmap batching rules under mixed precision, I think we should increase code coverage of this specific path. What do you think?

I'm gonna keep working on this and propose PRs with fixes and more tests. Any advice is very welcome!

### Versions

<details><summary>Environment details</summary>

PyTorch version: 2.14.0a0+gitb13b4f9
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 4.4.0
Libc version: glibc-2.35

Python version: 3.13.14 (main, Jul 23 2026, 14:44:45) [Clang 22.1.3 ] (64-bit runtime)
Python platform: Linux-6.8.0-136-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.6.20
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1080
Nvidia driver version: 580.173.02
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           39 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  12
On-line CPU(s) list:                     0-11
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
CPU family:                              6
Model:                                   158
Thread(s) per core:                      2
Core(s) per socket:                      6
Socket(s):                               1
Stepping:                                10
CPU max MHz:                             4100,0000
CPU min MHz:                             800,0000
BogoMIPS:                                4399.99
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb pti ssbd ibrs ibpb stibp tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp vnmi md_clear flush_l1d arch_capabilities ibpb_exit_to_user
Virtualization:                          VT-x
L1d cache:                               192 KiB (6 instances)
L1i cache:                               192 KiB (6 instances)
L2 cache:                                1,5 MiB (6 instances)
L3 cache:                                9 MiB (1 instance)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-11
Vulnerability Gather data sampling:      Mitigation; Microcode
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             KVM: Mitigation: VMX disabled
Vulnerability L1tf:                      Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds:                       Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Meltdown:                  Mitigation; PTI
Vulnerability Mmio stale data:           Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Mitigation; IBRS
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; IBRS; IBPB conditional; STIBP conditional; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Mitigation; Microcode
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] intel-cmplr-lib-ur==2026.1.1
[pip3] intel-openmp==2026.1.1
[pip3] mkl-include==2026.1.0
[pip3] mkl-static==2026.1.0
[pip3] numpy==2.5.1
[pip3] onemkl-license==2026.1.0
[pip3] optree==0.19.1
[pip3] tbb==2023.1.0
[pip3] tbb-devel==2023.1.0
[pip3] tcmlib==1.5.0
[pip3] torch==2.14.0a0+gitb13b4f9
[pip3] torchjd==0.17.0
[pip3] umf==1.1.0
[conda] Could not collect

</details>

cc @zou3519 @mcarilli @ptrblck @leslie-fang-intel @jgong5 @Chillee @samdow @kshitij12345
